### ISSUE TITLE
Shaking skeletons out of the closet

### ISSUE THREAD

**Issue by lpechacek** [OpenOrienteering/mapper#1773]
Let this ticket be a tracker for the random code problems found during debugging the Link Time Optimization (LTO) mystery. Feel free to analyze and fix anything reported here without closing the ticket itself. File separate tickets for the individual aspects if needed.

Situation: Mapper and its tests segfault when compiled with LTO on Fedora 33.  The selection of the compiler (GCC 10 / Clang 11) does not matter. I've observed similar effects also on openSUSE Tumbleweed (TW). Valgrind reported even more possible fixes on TW.

### First observation

On TW `QObject::connect: signal not found in Adwaita::Style` message from `file_format_t`, segfault at the end of test, warning from Valgrind about uninitialized variable at `src/gdal/ogr_file_format.cpp:1289`.

```
lpechacek@fmn:/tmp/lto-segv-debug> valgrind test/file_format_t
==7327== Memcheck, a memory error detector
==7327== Copyright (C) 2002-2017, and GNU GPL'd, by Julian Seward et al.
==7327== Using Valgrind-3.16.1 and LibVEX; rerun with -h for copyright info
==7327== Command: test/file_format_t
==7327== 
QObject::connect: signal not found in Adwaita::Style
QObject::connect: signal not found in Adwaita::Style
********* Start testing of FileFormatTest *********
Config: Using QtTest library 5.15.1, Qt 5.15.1 (x86_64-little_endian-lp64 shared (dynamic) release build; by GCC 10.2.1 20200825 [revision c0746a1beb1ba073c7981eb09f55b3d993b32e5c])
PASS   : FileFormatTest::initTestCase()
PASS   : FileFormatTest::mapCoordtoString(max,-1/0)
PASS   : FileFormatTest::mapCoordtoString(-2,max/1)
PASS   : FileFormatTest::mapCoordtoString(min,min/255)
PASS   : FileFormatTest::understandsTest(XML < xml legacy)
PASS   : FileFormatTest::understandsTest(XML < xml regular)
PASS   : FileFormatTest::understandsTest(XML < xml start)
PASS   : FileFormatTest::understandsTest(XML < xml incomplete 1)
PASS   : FileFormatTest::understandsTest(XML < xml incomplete 2)
PASS   : FileFormatTest::understandsTest(XML < '')
PASS   : FileFormatTest::understandsTest(XML < xml other)
PASS   : FileFormatTest::understandsTest(XML < 'OMAPxxx')
PASS   : FileFormatTest::understandsTest(XML < 0x0CADxxx)
PASS   : FileFormatTest::understandsTest(OCD < 0x0CADxxx)
PASS   : FileFormatTest::understandsTest(OCD < 0x0CAD)
PASS   : FileFormatTest::understandsTest(OCD < 0x0c)
PASS   : FileFormatTest::understandsTest(OCD < '')
PASS   : FileFormatTest::understandsTest(OCD < 'OMAPxxx')
PASS   : FileFormatTest::understandsTest(OCD < xml start)
PASS   : FileFormatTest::formatForDataTest(XML < xml legacy)
PASS   : FileFormatTest::formatForDataTest(XML < xml regular)
PASS   : FileFormatTest::formatForDataTest(XML < xml start)
PASS   : FileFormatTest::formatForDataTest(XML < xml incomplete 1)
PASS   : FileFormatTest::formatForDataTest(XML < xml incomplete 2)
PASS   : FileFormatTest::formatForDataTest(XML < '')
PASS   : FileFormatTest::formatForDataTest(XML < xml other)
PASS   : FileFormatTest::formatForDataTest(XML < 'OMAPxxx')
PASS   : FileFormatTest::formatForDataTest(XML < 0x0CADxxx)
PASS   : FileFormatTest::formatForDataTest(OCD < 0x0CADxxx)
PASS   : FileFormatTest::formatForDataTest(OCD < 0x0CAD)
PASS   : FileFormatTest::formatForDataTest(OCD < 0x0c)
PASS   : FileFormatTest::formatForDataTest(OCD < '')
PASS   : FileFormatTest::formatForDataTest(OCD < 'OMAPxxx')
PASS   : FileFormatTest::formatForDataTest(OCD < xml start)
ERROR 4: map.osm: No such file or directory
PASS   : FileFormatTest::issue_513_high_coordinates(testdata:issue-513-coords-outside-printable.xmap)
ERROR 4: map.osm: No such file or directory
PASS   : FileFormatTest::issue_513_high_coordinates(testdata:issue-513-coords-outside-printable.omap)
ERROR 4: map.osm: No such file or directory
PASS   : FileFormatTest::issue_513_high_coordinates(testdata:issue-513-coords-outside-qint32.omap)
PASS   : FileFormatTest::saveAndLoad(data:barrier.omap <> XML)
PASS   : FileFormatTest::saveAndLoad(data:rotated.omap <> XML)
PASS   : FileFormatTest::saveAndLoad(data:tags.omap <> XML)
PASS   : FileFormatTest::saveAndLoad(data:text-object.omap <> XML)
PASS   : FileFormatTest::saveAndLoad(data:undo.omap <> XML)
PASS   : FileFormatTest::saveAndLoad(data:/examples/complete map.omap <> XML)
PASS   : FileFormatTest::saveAndLoad(data:/examples/forest sample.omap <> XML)
PASS   : FileFormatTest::saveAndLoad(data:/examples/overprinting.omap <> XML)
PASS   : FileFormatTest::saveAndLoad(testdata:templates/world-file.xmap <> XML)
PASS   : FileFormatTest::saveAndLoad(data:/examples/complete map.omap <> OCD8)
PASS   : FileFormatTest::saveAndLoad(data:/examples/complete map.omap <> OCD9)
PASS   : FileFormatTest::saveAndLoad(data:/examples/complete map.omap <> OCD10)
PASS   : FileFormatTest::saveAndLoad(data:/examples/complete map.omap <> OCD11)
PASS   : FileFormatTest::saveAndLoad(data:/examples/complete map.omap <> OCD12)
PASS   : FileFormatTest::saveAndLoad(data:/examples/complete map.omap <> OCD (default))
PASS   : FileFormatTest::saveAndLoad(data:/examples/forest sample.omap <> OCD8)
PASS   : FileFormatTest::saveAndLoad(data:/examples/forest sample.omap <> OCD9)
PASS   : FileFormatTest::saveAndLoad(data:/examples/forest sample.omap <> OCD10)
PASS   : FileFormatTest::saveAndLoad(data:/examples/forest sample.omap <> OCD11)
PASS   : FileFormatTest::saveAndLoad(data:/examples/forest sample.omap <> OCD12)
PASS   : FileFormatTest::saveAndLoad(data:/examples/forest sample.omap <> OCD (default))
PASS   : FileFormatTest::saveAndLoad(data:/examples/overprinting.omap <> OCD8)
PASS   : FileFormatTest::saveAndLoad(data:/examples/overprinting.omap <> OCD9)
PASS   : FileFormatTest::saveAndLoad(data:/examples/overprinting.omap <> OCD10)
PASS   : FileFormatTest::saveAndLoad(data:/examples/overprinting.omap <> OCD11)
PASS   : FileFormatTest::saveAndLoad(data:/examples/overprinting.omap <> OCD12)
PASS   : FileFormatTest::saveAndLoad(data:/examples/overprinting.omap <> OCD (default))
PASS   : FileFormatTest::saveAndLoad(testdata:templates/world-file.xmap <> OCD8)
PASS   : FileFormatTest::saveAndLoad(testdata:templates/world-file.xmap <> OCD9)
PASS   : FileFormatTest::saveAndLoad(testdata:templates/world-file.xmap <> OCD10)
PASS   : FileFormatTest::saveAndLoad(testdata:templates/world-file.xmap <> OCD11)
PASS   : FileFormatTest::saveAndLoad(testdata:templates/world-file.xmap <> OCD12)
PASS   : FileFormatTest::saveAndLoad(testdata:templates/world-file.xmap <> OCD (default))
PASS   : FileFormatTest::pristineMapTest()
ERROR 6: Field of name 'Name' is not supported in GPX schema. Use GPX_USE_EXTENSIONS creation option to allow use of the <extensions> element.
==7327== Conditional jump or move depends on uninitialised value(s)
==7327==    at 0xFC50D8D: ??? (in /memfd:sljit (deleted))
==7327==    by 0x1104AE67: ???
==7327== 
==7327== Conditional jump or move depends on uninitialised value(s)
==7327==    at 0xFC50D8D: ??? (in /memfd:sljit (deleted))
==7327==    by 0x10D26147: ???
==7327== 
==7327== Conditional jump or move depends on uninitialised value(s)
==7327==    at 0xFC50D8D: ??? (in /memfd:sljit (deleted))
==7327==    by 0xF5E9FD7: ???
==7327== 
ERROR 6: Field of name 'Name' is not supported in GPX schema. Use GPX_USE_EXTENSIONS creation option to allow use of the <extensions> element.
ERROR 6: Geometry type of `Polygon' not supported in GPX.

==7327== Conditional jump or move depends on uninitialised value(s)
==7327==    at 0x6AA45B: setSRS (ogr_file_format.cpp:1289)
==7327==    by 0x6AA45B: importFeature (ogr_file_format.cpp:1031)
==7327==    by 0x6AA45B: importLayer (ogr_file_format.cpp:1024)
==7327==    by 0x6AA45B: OpenOrienteering::OgrFileImport::importImplementation() (ogr_file_format.cpp:842)
==7327==    by 0x4C2C5E: OpenOrienteering::Importer::doImport() (file_import_export.cpp:119)
==7327==    by 0x446C76: UnknownInlinedFun (file_format_t.cpp:930)
==7327==    by 0x446C76: FileFormatTest::qt_static_metacall(QObject*, QMetaObject::Call, int, void**) (moc_file_format_t.cpp:131)
==7327==    by 0x751B7C2: QMetaMethod::invoke(QObject*, Qt::ConnectionType, QGenericReturnArgument, QGenericArgument, QGenericArgument, QGenericArgument, QGenericArgument, QGenericArgument, QGenericArgument, QGenericArgument, QGenericArgument, QGenericArgument, QGenericArgument) const (in /usr/lib64/libQt5Core.so.5.15.1)
==7327==    by 0x4BACF12: ??? (in /usr/lib64/libQt5Test.so.5.15.1)
==7327==    by 0x4BAD7AB: ??? (in /usr/lib64/libQt5Test.so.5.15.1)
==7327==    by 0x4BADD23: ??? (in /usr/lib64/libQt5Test.so.5.15.1)
==7327==    by 0x4BAE1CD: QTest::qRun() (in /usr/lib64/libQt5Test.so.5.15.1)
==7327==    by 0x4BAE56B: QTest::qExec(QObject*, int, char**) (in /usr/lib64/libQt5Test.so.5.15.1)
==7327==    by 0x452F5F: main (file_format_t.cpp:955)
==7327== 
PASS   : FileFormatTest::ogrExportTest(complete map)
PASS   : FileFormatTest::cleanupTestCase()
Totals: 73 passed, 0 failed, 0 skipped, 0 blacklisted, 126435ms
********* Finished testing of FileFormatTest *********
==7327== Thread 2 QDBusConnectionM:
==7327== Invalid read of size 8
==7327==    at 0x753B849: QObject::disconnect(QObject const*, char const*, QObject const*, char const*) (in /usr/lib64/libQt5Core.so.5.15.1)
==7327==    by 0xD64C6BF: ??? (in /usr/lib64/libQt5DBus.so.5.15.1)
==7327==    by 0xD640A94: ??? (in /usr/lib64/libQt5DBus.so.5.15.1)
==7327==    by 0x732D910: ??? (in /usr/lib64/libQt5Core.so.5.15.1)
==7327==    by 0x7F66EB0: start_thread (in /lib64/libpthread-2.32.so)
==7327==    by 0x7C32CCE: clone (in /lib64/libc-2.32.so)
==7327==  Address 0xb52f7c0 is 0 bytes inside a block of size 120 free'd
==7327==    at 0x483AEAB: operator delete(void*) (in /usr/lib64/valgrind/vgpreload_memcheck-amd64-linux.so)
==7327==    by 0x4D61E05: QApplication::~QApplication() (in /usr/lib64/libQt5Widgets.so.5.15.1)
==7327==    by 0x452F73: main (file_format_t.cpp:955)
==7327==  Block was alloc'd at
==7327==    at 0x4839DEF: operator new(unsigned long) (in /usr/lib64/valgrind/vgpreload_memcheck-amd64-linux.so)
==7327==    by 0xD5E5ED7: Adwaita::StylePlugin::create(QString const&) (in /usr/lib64/qt5/plugins/styles/adwaita.so)
==7327==    by 0x4DD0029: QStyleFactory::create(QString const&) (in /usr/lib64/libQt5Widgets.so.5.15.1)
==7327==    by 0x4D644D0: QApplication::style() (in /usr/lib64/libQt5Widgets.so.5.15.1)
==7327==    by 0x4D64A49: QApplicationPrivate::initialize() (in /usr/lib64/libQt5Widgets.so.5.15.1)
==7327==    by 0x4D65F45: QApplicationPrivate::init() (in /usr/lib64/libQt5Widgets.so.5.15.1)
==7327==    by 0x452F20: main (file_format_t.cpp:955)
==7327== 
==7327== Invalid read of size 8
==7327==    at 0x7535C10: QObject::metaObject() const (in /usr/lib64/libQt5Core.so.5.15.1)
==7327==    by 0x753B84D: QObject::disconnect(QObject const*, char const*, QObject const*, char const*) (in /usr/lib64/libQt5Core.so.5.15.1)
==7327==    by 0xD64C6BF: ??? (in /usr/lib64/libQt5DBus.so.5.15.1)
==7327==    by 0xD640A94: ??? (in /usr/lib64/libQt5DBus.so.5.15.1)
==7327==    by 0x732D910: ??? (in /usr/lib64/libQt5Core.so.5.15.1)
==7327==    by 0x7F66EB0: start_thread (in /lib64/libpthread-2.32.so)
==7327==    by 0x7C32CCE: clone (in /lib64/libc-2.32.so)
==7327==  Address 0xb52f7c8 is 8 bytes inside a block of size 120 free'd
==7327==    at 0x483AEAB: operator delete(void*) (in /usr/lib64/valgrind/vgpreload_memcheck-amd64-linux.so)
==7327==    by 0x4D61E05: QApplication::~QApplication() (in /usr/lib64/libQt5Widgets.so.5.15.1)
==7327==    by 0x452F73: main (file_format_t.cpp:955)
==7327==  Block was alloc'd at
==7327==    at 0x4839DEF: operator new(unsigned long) (in /usr/lib64/valgrind/vgpreload_memcheck-amd64-linux.so)
==7327==    by 0xD5E5ED7: Adwaita::StylePlugin::create(QString const&) (in /usr/lib64/qt5/plugins/styles/adwaita.so)
==7327==    by 0x4DD0029: QStyleFactory::create(QString const&) (in /usr/lib64/libQt5Widgets.so.5.15.1)
==7327==    by 0x4D644D0: QApplication::style() (in /usr/lib64/libQt5Widgets.so.5.15.1)
==7327==    by 0x4D64A49: QApplicationPrivate::initialize() (in /usr/lib64/libQt5Widgets.so.5.15.1)
==7327==    by 0x4D65F45: QApplicationPrivate::init() (in /usr/lib64/libQt5Widgets.so.5.15.1)
==7327==    by 0x452F20: main (file_format_t.cpp:955)
==7327== 
==7327== Invalid read of size 8
==7327==    at 0x7535C14: QObject::metaObject() const (in /usr/lib64/libQt5Core.so.5.15.1)
==7327==    by 0x753B84D: QObject::disconnect(QObject const*, char const*, QObject const*, char const*) (in /usr/lib64/libQt5Core.so.5.15.1)
==7327==    by 0xD64C6BF: ??? (in /usr/lib64/libQt5DBus.so.5.15.1)
==7327==    by 0xD640A94: ??? (in /usr/lib64/libQt5DBus.so.5.15.1)
==7327==    by 0x732D910: ??? (in /usr/lib64/libQt5Core.so.5.15.1)
==7327==    by 0x7F66EB0: start_thread (in /lib64/libpthread-2.32.so)
==7327==    by 0x7C32CCE: clone (in /lib64/libc-2.32.so)
==7327==  Address 0xb52f8a8 is 40 bytes inside a block of size 184 free'd
==7327==    at 0x483B08B: operator delete(void*, unsigned long) (in /usr/lib64/valgrind/vgpreload_memcheck-amd64-linux.so)
==7327==    by 0xD5C7378: Adwaita::Style::~Style() (in /usr/lib64/qt5/plugins/styles/adwaita.so)
==7327==    by 0x4D61E05: QApplication::~QApplication() (in /usr/lib64/libQt5Widgets.so.5.15.1)
==7327==    by 0x452F73: main (file_format_t.cpp:955)
==7327==  Block was alloc'd at
==7327==    at 0x4839DEF: operator new(unsigned long) (in /usr/lib64/valgrind/vgpreload_memcheck-amd64-linux.so)
==7327==    by 0x4DD7902: QCommonStyle::QCommonStyle() (in /usr/lib64/libQt5Widgets.so.5.15.1)
==7327==    by 0xD5CE1CD: Adwaita::Style::Style(bool) (in /usr/lib64/qt5/plugins/styles/adwaita.so)
==7327==    by 0xD5E5EE7: Adwaita::StylePlugin::create(QString const&) (in /usr/lib64/qt5/plugins/styles/adwaita.so)
==7327==    by 0x4DD0029: QStyleFactory::create(QString const&) (in /usr/lib64/libQt5Widgets.so.5.15.1)
==7327==    by 0x4D644D0: QApplication::style() (in /usr/lib64/libQt5Widgets.so.5.15.1)
==7327==    by 0x4D64A49: QApplicationPrivate::initialize() (in /usr/lib64/libQt5Widgets.so.5.15.1)
==7327==    by 0x4D65F45: QApplicationPrivate::init() (in /usr/lib64/libQt5Widgets.so.5.15.1)
==7327==    by 0x452F20: main (file_format_t.cpp:955)
==7327== 
==7327== Invalid read of size 8
==7327==    at 0x753B156: ??? (in /usr/lib64/libQt5Core.so.5.15.1)
==7327==    by 0x753BC78: QObject::disconnect(QObject const*, char const*, QObject const*, char const*) (in /usr/lib64/libQt5Core.so.5.15.1)
==7327==    by 0xD64C6BF: ??? (in /usr/lib64/libQt5DBus.so.5.15.1)
==7327==    by 0xD640A94: ??? (in /usr/lib64/libQt5DBus.so.5.15.1)
==7327==    by 0x732D910: ??? (in /usr/lib64/libQt5Core.so.5.15.1)
==7327==    by 0x7F66EB0: start_thread (in /lib64/libpthread-2.32.so)
==7327==    by 0x7C32CCE: clone (in /lib64/libc-2.32.so)
==7327==  Address 0xb52f7c8 is 8 bytes inside a block of size 120 free'd
==7327==    at 0x483AEAB: operator delete(void*) (in /usr/lib64/valgrind/vgpreload_memcheck-amd64-linux.so)
==7327==    by 0x4D61E05: QApplication::~QApplication() (in /usr/lib64/libQt5Widgets.so.5.15.1)
==7327==    by 0x452F73: main (file_format_t.cpp:955)
==7327==  Block was alloc'd at
==7327==    at 0x4839DEF: operator new(unsigned long) (in /usr/lib64/valgrind/vgpreload_memcheck-amd64-linux.so)
==7327==    by 0xD5E5ED7: Adwaita::StylePlugin::create(QString const&) (in /usr/lib64/qt5/plugins/styles/adwaita.so)
==7327==    by 0x4DD0029: QStyleFactory::create(QString const&) (in /usr/lib64/libQt5Widgets.so.5.15.1)
==7327==    by 0x4D644D0: QApplication::style() (in /usr/lib64/libQt5Widgets.so.5.15.1)
==7327==    by 0x4D64A49: QApplicationPrivate::initialize() (in /usr/lib64/libQt5Widgets.so.5.15.1)
==7327==    by 0x4D65F45: QApplicationPrivate::init() (in /usr/lib64/libQt5Widgets.so.5.15.1)
==7327==    by 0x452F20: main (file_format_t.cpp:955)
==7327== 
==7327== Invalid read of size 8
==7327==    at 0x753B15A: ??? (in /usr/lib64/libQt5Core.so.5.15.1)
==7327==    by 0x753BC78: QObject::disconnect(QObject const*, char const*, QObject const*, char const*) (in /usr/lib64/libQt5Core.so.5.15.1)
==7327==    by 0xD64C6BF: ??? (in /usr/lib64/libQt5DBus.so.5.15.1)
==7327==    by 0xD640A94: ??? (in /usr/lib64/libQt5DBus.so.5.15.1)
==7327==    by 0x732D910: ??? (in /usr/lib64/libQt5Core.so.5.15.1)
==7327==    by 0x7F66EB0: start_thread (in /lib64/libpthread-2.32.so)
==7327==    by 0x7C32CCE: clone (in /lib64/libc-2.32.so)
==7327==  Address 0xb52f8c0 is 64 bytes inside a block of size 184 free'd
==7327==    at 0x483B08B: operator delete(void*, unsigned long) (in /usr/lib64/valgrind/vgpreload_memcheck-amd64-linux.so)
==7327==    by 0xD5C7378: Adwaita::Style::~Style() (in /usr/lib64/qt5/plugins/styles/adwaita.so)
==7327==    by 0x4D61E05: QApplication::~QApplication() (in /usr/lib64/libQt5Widgets.so.5.15.1)
==7327==    by 0x452F73: main (file_format_t.cpp:955)
==7327==  Block was alloc'd at
==7327==    at 0x4839DEF: operator new(unsigned long) (in /usr/lib64/valgrind/vgpreload_memcheck-amd64-linux.so)
==7327==    by 0x4DD7902: QCommonStyle::QCommonStyle() (in /usr/lib64/libQt5Widgets.so.5.15.1)
==7327==    by 0xD5CE1CD: Adwaita::Style::Style(bool) (in /usr/lib64/qt5/plugins/styles/adwaita.so)
==7327==    by 0xD5E5EE7: Adwaita::StylePlugin::create(QString const&) (in /usr/lib64/qt5/plugins/styles/adwaita.so)
==7327==    by 0x4DD0029: QStyleFactory::create(QString const&) (in /usr/lib64/libQt5Widgets.so.5.15.1)
==7327==    by 0x4D644D0: QApplication::style() (in /usr/lib64/libQt5Widgets.so.5.15.1)
==7327==    by 0x4D64A49: QApplicationPrivate::initialize() (in /usr/lib64/libQt5Widgets.so.5.15.1)
==7327==    by 0x4D65F45: QApplicationPrivate::init() (in /usr/lib64/libQt5Widgets.so.5.15.1)
==7327==    by 0x452F20: main (file_format_t.cpp:955)
==7327== 
==7327== 
==7327== HEAP SUMMARY:
==7327==     in use at exit: 751,848 bytes in 4,592 blocks
==7327==   total heap usage: 7,360,264 allocs, 7,355,672 frees, 1,831,140,039 bytes allocated
==7327== 
==7327== LEAK SUMMARY:
==7327==    definitely lost: 1,880 bytes in 47 blocks
==7327==    indirectly lost: 2,350 bytes in 47 blocks
==7327==      possibly lost: 160 bytes in 1 blocks
==7327==    still reachable: 747,434 bytes in 4,496 blocks
==7327==         suppressed: 24 bytes in 1 blocks
==7327== Rerun with --leak-check=full to see details of leaked memory
==7327== 
==7327== Use --track-origins=yes to see where uninitialised values come from
==7327== For lists of detected and suppressed errors, rerun with: -s
==7327== ERROR SUMMARY: 9 errors from 9 contexts (suppressed: 0 from 0)
```

**Comment by lpechacek** [comment]
Reproducer. Adjust the paths after `-S` and `-B` to your liking.
```
$ CXX=clang++ CC=clang /usr/bin/cmake -S ~/mapper/ -B /tmp/lto-segv-debug -DCMAKE_C_FLAGS_RELEASE:STRING=-DNDEBUG -DCMAKE_CXX_FLAGS_RELEASE:STRING=-DNDEBUG -DCMAKE_VERBOSE_MAKEFILE:BOOL=ON -DCMAKE_BUILD_TYPE=Release -DMapper_WITH_COVE:BOOL=OFF -DCMAKE_CXX_FLAGS=-flto -DCMAKE_C_FLAGS=-flto -DCMAKE_EXE_LINKER_FLAGS=-flto
$ cd /tmp/lto-segv-debug/
$ make -j8 -C test/ file_format_t && test/file_format_t
```

In TW it causes the warning about signal connections. In F33 it triggers the early crash.

**Comment by dg0yt** [comment]
Thanks for looking at this. I'm still concerned about seeing a crash so early in `main`, during construction of `QtSingleApplication`. At this point, there is little what could be wrong on our side. How about testing without `QtSingleApplication` (by undefining it), or trying a minimal example with `QtSingleApplication`?

**Comment by dg0yt** [comment]
Looking for LTO bugreports:
https://bugreports.qt.io/browse/QTBUG-86173,
with a link to a similar report: https://github.com/clearlinux/distribution/issues/2169

**Comment by lpechacek** [comment]
Another data point. With GCC 10, remove CoVe (`-DMapper_WITH_COVE:BOOL=OFF` option to `cmake`) and remove OCD8 legacy code (commit c6168807) from v0.9.4, and the segfault is gone even with LTO enabled. Adding either of them back to the mix re-introduces the crash.

> How about testing without `QtSingleApplication` (by undefining it), or trying a minimal example with `QtSingleApplication`?

Please mind that it's only Mapper that has `QtSingleApplication`'s methods on the stack. The tests fail with a different stack trace. One pattern among the tests is that tests featuring `QTEST_GUILESS_MAIN()` uniformly pass. Those featuring `QTEST_GUILESS_MAIN()` mostly crash, with the exception of `style_t` and `sensors_t`.

Another observation, there is a possibly false warning from GCC talking about unintialized variable:
```
/home/lpechacek/mapper/src/gui/map/map_editor.cpp: In member function ‘connectPathsClicked’:
/home/lpechacek/mapper/src/gui/map/map_editor.cpp:3183:40: warning: ‘best_object_b_index’ may be used uninitialized in this function [-Wmaybe-uninitialized]
 3183 |   if (!undo_objects[best_object_b_index])
      |                                        ^
/home/lpechacek/mapper/src/gui/map/map_editor.cpp:3126:7: note: ‘best_object_b_index’ was declared here
 3126 |   int best_object_b_index = 0;
      |       ^
/home/lpechacek/mapper/src/gui/map/map_editor.cpp:3167:20: warning: ‘path_part_b’ may be used uninitialized in this function [-Wmaybe-uninitialized]
 3167 |        best_part_b = path_part_b;
      |                    ^
/home/lpechacek/mapper/src/gui/map/map_editor.cpp:3151:10: note: ‘path_part_b’ was declared here
 3151 |      int path_part_b;
      |          ^
/home/lpechacek/mapper/src/gui/map/map_editor.cpp:3199:31: warning: ‘best_part_b_begin’ may be used uninitialized in this function [-Wmaybe-uninitialized]
 3199 |   else if (!best_part_a_begin && best_part_b_begin)
      |                               ^
/home/lpechacek/mapper/src/gui/map/map_editor.cpp:3130:8: note: ‘best_part_b_begin’ was declared here
 3130 |   bool best_part_b_begin = false;
      |        ^
```
This strongly correlates with the crash.

At this point, I think it is some combination of libraries and/or code snippets that either confuse LTO or violate some obvious rule.

**Comment by dg0yt** [comment]
> Please mind that it's only Mapper that has `QtSingleApplication`'s methods on the stack. The tests fail with a different stack trace.

Yes, I realized that after posting.

> At this point, I think it is some combination of libraries and/or code snippets that either confuse LTO or violate some obvious rule.

See the QTBUG. Takes some time to read it all. But includes everything: Fedora, Tumbleweed, LTO, QApplication, CMake.

**Comment by lpechacek** [comment]
> See the QTBUG. Takes some time to read it all. But includes everything: Fedora, Tumbleweed, LTO, QApplication, CMake.

Yes, just started. It indeed looks relevant. I'll give it more time tomorrow. Thanks!

**Comment by lpechacek** [comment]
I've read through the Qt bug and the related openSUSE report. These two contain the most information. The topic is quite complex, and I'm unfamiliar with some of the basic concepts there. Still, from what I understood, there is some unwelcome symbol mismatch during the link phase, demonstrating itself as a null pointer dereference. I can see the same in the Fedora 33 build, so for the moment, I'd say we are hitting the same issue. The numerous remedies described there fix the issue. My takeaway is that Mapper should not adapt to the build system, but instead, the toolchain should produce the correct output. openSUSE guys (Fabian, Michael, Martin, and Richi) actively work on the fix. Fedora guys have been informed through https://bugzilla.redhat.com/show_bug.cgi?id=1900527.

Regarding the Mapper skeletons in the closet, there are rare. :) I ran all Mapper tests through Valgrind and I've placed #1782 for the only bug I found. The other Valgrind findings apply either to the linked libraries (e.g. the `memfd:sljit` errors originate in `pcre2`) or look like false positives. I'm done with this code audit round and I'm again surprised how robust is the implementation.
